### PR TITLE
inkscape: allow loading external extensions

### PR DESCRIPTION
**Pull request by raboof**
###### Motivation for this change

My longer-term motivation is that I want to be able to load inkcut as an
extension, but let's start with getting the infrastructure ready and adding
a simple one.

I added the 'hexmap' extension to show the process. You could use this for
example like:

```
{ pkgs ? import <nixpkgs> {} }:

pkgs.mkShell {
  buildInputs = [
    (pkgs.inkscape-with-extensions.override {
      inkscapeExtensions = [
        pkgs.inkscape-extensions.hexmap
      ];
    })
  ];
}
```

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

/cc @jtojnar

### CODE REVIEW

File: pkgs/applications/graphics/inkscape/extensions.nix
Comment (jtojnar) [inline pkgs/applications/graphics/inkscape/extensions.nix:20]: When overriding `installPhase`, you should also provide `runHook preInstall` and `runHook postInstall`. Though, in this case, it is better to rely on the default `installPhase`.

File: pkgs/top-level/all-packages.nix
Comment (symphorien) [inline pkgs/top-level/all-packages.nix:20830]: since this file evaluates as a set of derivations:
```suggestion
  inkscape-extensions = callPackages ../applications/graphics/inkscape/extensions.nix {};
```

**Comment by symphorien** [comment]
I could run inkscape with hexmap and it works. The `inkscape-with-extensions` attrset probably needs `recurseIntoAttrs` for discoverability.

**Comment by raboof** [comment]
> I could run inkscape with hexmap and it works.

> The inkscape-with-extensions attrset probably needs recurseIntoAttrs for discoverability.

Nice, I didn't know about that! Tested with `nix search -u hexmap` and indeed adding `recurseIntoAttrs` makes it show up.

**Review by symphorien — APPROVED** [review]

**Comment by risicle** [comment]
This does work on macos (tested 10.14), just with a slight quirk...

Screenshot 2020-09-08 at 23 18 39

It's far from a showstopper and I wouldn't know where to start fixing it.

I like this - think it's nearly merge-o-clock?

**Comment by raboof** [comment]
> my `inkscape-with-extensions` doesn't have links to the hexmap files in its extensions directory, on both linux and macos

Did you add it to the `inkscapeExtensions` attribute? I followed the convention by vscode to have this list empty by default, you only get the extensions that you add yourself.

> could you not have used `symlinkJoin` from `pkgs/build-support/trivial-builders.nix` instead of writing your own linking loop?

Possibly! I'll give it a try later.

**Comment by raboof** [comment]
> Also, could you not have used symlinkJoin from pkgs/build-support/trivial-builders.nix instead of writing your own linking loop?

Yes, very nice simplification - done, thanks!

**Comment by risicle** [comment]
I think the last version you pushed is a little broken: https://gist.github.com/GrahamcOfBorg/b268ea1ccbe05077fa97f8288a740dd5

**Comment by risicle** [comment]
> Did you add it to the inkscapeExtensions attribute? I followed the convention by vscode to have this list empty by default, you only get the extensions that you add yourself.

Ok - so I guess the intended way to use this would be `inkscape-with-extensions.override { inkscapeExtensions = [ inkscape-extensions.hexmap ]; }`?

Do you think there might be value in having something like an `inkscapeFull` with _all_ known extensions to make this available in a more... ergonomic way? i.e. possible to address through a single attribute, not having to break into expression-writing.

(I also came across how `terraform` handles extensions today, which allows you to do `terraform_0_12.withPlugins (p: p.fooplugin)`, which is pretty neat)

**Comment by raboof** [comment]
> > Did you add it to the inkscapeExtensions attribute? I followed the convention by vscode to have this list empty by default, you only get the extensions that you add yourself.
> 
> Ok - so I guess the intended way to use this would be `inkscape-with-extensions.override { inkscapeExtensions = [ inkscape-extensions.hexmap ]; }`?

Indeed! I was following the lead of e.g. vscode-with-extensions and dde-dock-with-plugins.

> Do you think there might be value in having something like an `inkscapeFull` with _all_ known extensions to make this available in a more... ergonomic way? i.e. possible to address through a single attribute, not having to break into expression-writing.

Sounds good, yeah - though that could get 'heavy' (with some extensions pulling in further dependencies), so I guess it'd be good to have both options?

> (I also came across how `terraform` handles extensions today, which allows you to do `terraform_0_12.withPlugins (p: p.fooplugin)`, which is pretty neat)

I agree that does look quite neat.

**Review by risicle — APPROVED** [review]
I can confirm this WFM on non-nixos linux x86_64 & macos 10.14.

**Comment by risicle** [comment]
Regarding an `inkscapeFull`, I realize it could get a bit heavy but I think without it or something similar we'd be damning some of our less technical (or simply newer) users from being able to access these at all.

Anyway, I see you're looking to do more work around inkscape extensions in the future, so I'll leave it to you to consider it.

**Comment by risicle** [comment]
Oh one further thing - would you add yourself as maintainer of the extension?

**Comment by raboof** [comment]
> I realize it could get a bit heavy but I think without it or something similar we'd be damning some of our less technical (or simply newer) users from being able to access these at all.

Yeah, I like the idea, but let's keep it for a later PR

> Oh one further thing - would you add yourself as maintainer of the extension?

Sure, done

**Pull request merged**
